pyomo/contrib/doe/measurements.py

```diff
@@ -25,328 +25,311 @@
+        all_variable_indices = list(itertools.product(*all_index_list))
+
+        # list store all names added this time
+        added_names = []
+        # iterate over index combinations ["CA", 1], ["CA", 2], ..., ["CC", 2], ["CC", 3]
+        for index_instance in all_variable_indices:
+            var_name_index_string = var_name + "["
+            for i, idx in enumerate(index_instance):
+                var_name_index_string += str(idx)
+
+                # if i is the last index, close the []. if not, add a "," for the next index.
+                if i == len(index_instance) - 1:
+                    var_name_index_string += "]"
+                else:
+                    var_name_index_string += ","
+
+            self.variable_names.append(var_name_index_string)
+            added_names.append(var_name_index_string)
+
+        return added_names
+
+    def _check_valid_input(
+        self,
+        len_indices,
+        var_name,
+        indices,
+        time_index_position,
+        values,
+        lower_bounds,
+        upper_bounds,
+    ):
+        """
+        Check if the measurement information provided are valid to use.
+        """
+        assert type(var_name) is str, "var_name should be a string."
+
+        if time_index_position not in indices:
+            raise ValueError("time index cannot be found in indices.")
+
+        # if given a list, check if bounds have the same length with flattened variable
+        if values and len(values) != len_indices:
+            raise ValueError("Values is of different length with indices.")
+
+        if (
+            lower_bounds
+            and type(lower_bounds) == list
+            and len(lower_bounds) != len_indices
+        ):
+            raise ValueError("Lowerbounds is of different length with indices.")
 
+        if (
+            upper_bounds
+            and type(upper_bounds) == list
+            and len(upper_bounds) != len_indices
+        ):
+            raise ValueError("Upperbounds is of different length with indices.")
+
+
+class MeasurementVariables(VariablesWithIndices):
+    def __init__(self):
         """
-        flatten_names = []
-        for j in measure_name_and_index.keys():
-            if measure_name_and_index[j] is not None:  # if it has extra index
-                for ind in measure_name_and_index[j]:
-                    flatten_name = j + self.ind_string + str(ind)
-                    flatten_names.append(flatten_name)
-            else:
-                flatten_names.append(j)
-
-        self.flatten_measure_name = flatten_names
-
-    def _generate_variance(self, flatten_measure_name, variance, name_and_index):
+        This class stores information on which algebraic and differential variables in the Pyomo model are considered measurements.
         """
-        Generate the variance dictionary
+        super().__init__()
+        self.variance = {}
 
+    def set_variable_name_list(self, self_define_res, variance=1):
```

Comment:
This function is to specify a list of measurement names given by the user. If users give this function an input like ["CA[0,1]", "CB[2,2]"], this will directly be the final variable names as they wish. Then it sets up the variance if user defines something other than None.